### PR TITLE
Introduce Aggregated Updating Of ASPA Table

### PR DESCRIPTION
**Pull request by carl-tud**
## Changes
This PR introduces a mechanism to reduce iterating over ASPA data (in the `aspa_array`). `aspa_table_update` receives an array of *add* and *remove* operations. Internally, this new function is able to detect conflicts in newly arrived ASPA data and can avoid accessing memory by resolving announcement/withdrawal pairs (e.g. `ASPA PDU (flag=announce) CAS 42 with providers ...`; `ASPA PDU (flag=withdraw) CAS 42`). 

Previously we didn't consider we need to release the various provider sets referenced by each record ⏤ This PR fixes that issue, too.

Additionally, `aspa_store`'s complexity was significantly reduced, a separate file + header therefore isn't necessary anymore.

## Tests
Tests in `tests/test_aspa.c` need to be rewritten, a template is commented out.
This PR's functionality was tested with `fishy-rtr-server` (I will push the changes I made to `rtrlab` into a separate branch once I've cleaned up the code)

### CODE REVIEW

File: rtrlib/aspa/aspa.h [hunk 1]
```diff
@@ -141,70 +122,78 @@ void aspa_table_init(struct aspa_table *aspa_table, aspa_update_fp update_fp);
...
- * @return aspa_ERROR On error.
- * @return aspa_DUPLICATE_RECORD If an identical aspa_record already exists
+ * @param[in] rtr_socket origin socket of the record
+ * @param notify A boolean value determining whether clients about removed records.
+ * @return @c ASPA_SUCCESS On success.
+ * @return @c ASPA_ERROR On error.
  */
-int aspa_table_add(struct aspa_table *aspa_table, struct aspa_record *aspa_record, struct rtr_socket *rtr_socket,
-		   bool replace);
+enum aspa_status aspa_table_src_remove(struct aspa_table *aspa_table, struct rtr_socket *rtr_socket, bool notify);
```
Comment (tanneberger) [inline rtrlib/aspa/aspa.h:132]: this is a weird name ... and it doesn't really match witht the docs above

File: rtrlib/aspa/aspa.h [hunk 2]
```diff
@@ -141,70 +122,78 @@ void aspa_table_init(struct aspa_table *aspa_table, aspa_update_fp update_fp);
...
+
+enum aspa_verification_result {
+	ASPA_AS_PATH_UNKNOWN,
+	ASPA_AS_PATH_INVALID,
+	ASPA_AS_PATH_VALID,
+};
 
 /**
- * @brief Removes aspa_record from aspa_table
+ * @brief Verifies an @c AS_PATH .
```
Comment (tanneberger) [inline rtrlib/aspa/aspa.h]: @c ?

File: rtrlib/aspa/aspa_array/aspa_array.c [hunk 3]
```diff
@@ -23,140 +23,222 @@ int aspa_array_create(struct aspa_array **vector_pointer)
...
-	size_t j = vector->size;
-
-	// we are moving from the back of the array to the front while looking
-	// for the place where to insert our new record
-	while (j > 0 && vector->data[j - 1].customer_asn > record.customer_asn) {
-		// moving data to the right
-		vector->data[j] = vector->data[j - 1];
-		j -= 1;
-	}
+enum aspa_status aspa_array_set_record(struct aspa_array *array, size_t index, struct aspa_record *record)
```
Comment (tanneberger) [inline rtrlib/aspa/aspa_array/aspa_array.c]: no bounds check ? / garbage collection ?

File: rtrlib/aspa/aspa_private.h [hunk 4]
```diff
@@ -19,11 +19,25 @@
...
 
+#include "../rtr/rtr.h"
 #include "aspa.h"
 
 #include <stdbool.h>
 #include <stdint.h>
 
+#define ASPA_UPDATE_IN_PLACE 1
+
+// MARK: - Storage
```
Comment (tanneberger) [inline rtrlib/aspa/aspa_private.h]: remove ?

File: rtrlib/aspa/aspa_private.h [hunk 5]
```diff
@@ -32,8 +46,89 @@
...
+
+/**
+ * @brief ASPA update finalization arguments
+ * @param unused_provider_arrays Array of pointers to now unused provider sets.
+ * @param unused_provider_array_len Number of unused provider sets.
+ */
+struct aspa_update_finalization_args {
+	uint32_t **unused_provider_arrays;
+	size_t unused_provider_array_len;
+#ifndef ASPA_UPDATE_IN_PLACE
```
Comment (tanneberger) [inline rtrlib/aspa/aspa_private.h:83]: ohhhh god - this is the direct pathway to circle 9 of hell.

File: rtrlib/aspa/aspa_verification.c [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+
+	pthread_rwlock_unlock(&aspa_table->lock);
+	return ASPA_AS_PATH_UNKNOWN;
+}
+
+RTRLIB_EXPORT enum aspa_verification_result aspa_verify_as_path(struct aspa_table *aspa_table, uint32_t as_path[],
+								size_t len, enum aspa_direction direction)
+{
+	switch (direction) {
+	case ASPA_UPSTREAM:
```
Comment (tanneberger) [inline rtrlib/aspa/aspa_verification.c:258]: identation

File: rtrlib/aspa/aspa_private.h (hunk 5 above)
Comment (carl-tud) [reply rtrlib/aspa/aspa_private.h:83]: do you mean using a macro here or the args struct itself? If it's really the macro bothering you, we will get rid of that once we figured out which option is faster during the interop testing. (Depending on these results, this struct is either going to be renamed to `aspa_update_cleanup_args` or remain like that.

File: rtrlib/aspa/aspa_verification.c (hunk 6 above)
Comment (carl-tud) [reply rtrlib/aspa/aspa_verification.c:258]: clang-format does that... + linux kernel coding style

File: rtrlib/aspa/aspa.h (hunk 1 above)
Comment (carl-tud) [reply rtrlib/aspa/aspa.h:132]: You're right, docs don't match; naming is kind of dictated by `spki_table` and `pfx_table`

File: rtrlib/aspa/aspa_array/aspa_array.c (hunk 3 above)
Comment (carl-tud) [reply rtrlib/aspa/aspa_array/aspa_array.c]: oh right I forgot

Comment (carl-tud) [reply rtrlib/aspa/aspa_array/aspa_array.c]: I know why.. `aspa_array_set_record` isn't called anywhere and can be removed

**Pull request merged**